sklearn/metrics/classification.py

```diff
@@ -266,17 +266,25 @@ def confusion_matrix(y_true, y_pred, labels=None, sample_weight=None):
     check_consistent_length(sample_weight, y_true, y_pred)
 
     n_labels = labels.size
-    label_to_ind = dict((y, x) for x, y in enumerate(labels))
-    # convert yt, yp into index
-    y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-    y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
-
-    # intersect y_pred, y_true with labels, eliminate items not in labels
-    ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
-    y_pred = y_pred[ind]
-    y_true = y_true[ind]
-    # also eliminate weights of eliminated items
-    sample_weight = sample_weight[ind]
+    # If labels are not consecitive integers starting from zero, then
+    # yt, yp must be converted into index form
+    need_index_conversion = not (
```

Comment:
Surely `np.all(labels == np.arange(len(labels))` is just as fast for a reasonable number of classes and much more readable?

Reply:
It is both faster and more readable, when I originally wrote this I didn't consider that the ordering of the labels was significant, so the `np.diff` was a quick fix after realizing this. 

```python
In [1]: labels = np.arange(100)

In [2]: %timeit labels.min() == 0 and np.all(np.diff(labels) == 1) 
8.3 µs ± 162 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [3]: %timeit np.all(np.arange(len(labels)) == labels)
3.6 µs ± 47.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```